File: 2133-check-if-every-row-and-column-contains-all-numbers/2133-check-if-every-row-and-column-contains-all-numbers.rb
# @param {Integer[][]} matrix
# @return {Boolean}
def check_single(row)
    tally = [0] * row.length
    (0...row.length).each do |i|
        if tally[row[i]-1] > 0
            return false
        else
            tally[row[i]-1] += 1
        end
    end
    return true
end

def check_valid(matrix)
    (0...matrix.length).each do |r|
        return false if !check_single(matrix[r])
    end
    
    (0...matrix.length).each do |c|
        return false if !check_single(matrix.map{ |r| r[c] })
    end
    return true
end

File: 2133-check-if-every-row-and-column-contains-all-numbers/2133-check-if-every-row-and-column-contains-all-numbers.js
/**
 * @param {number[][]} matrix
 * @return {boolean}
 */
var checkValid = function(matrix) {
    const checkSingle = row => {
        const tally = new Array(row.length).fill(0);
        for (let i = 0; i < row.length; i++) {
            if (tally[row[i] - 1] > 0) {
                return false;
            } else {
                tally[row[i] - 1]++;
            }
        }
        return true;
    };
    
    for (let r = 0; r < matrix.length; r++) {
        if (!checkSingle(matrix[r])) return false;
    }
    
    for (let c = 0; c < matrix.length; c++) {
        const col = matrix.map(r => r[c]);
        if (!checkSingle(col)) return false;
    }
    return true;
};